packages/opentelemetry-api/src/api/global-utils.ts

```diff
@@ -15,54 +15,74 @@
  */
 
 import { ContextManager } from '@opentelemetry/context-base';
+import * as semver from 'semver';
 import { TextMapPropagator } from '../context/propagation/TextMapPropagator';
 import { MeterProvider } from '../metrics/MeterProvider';
-import { TracerProvider } from '../trace/tracer_provider';
 import { _globalThis } from '../platform';
+import { TracerProvider } from '../trace/tracer_provider';
+import { VERSION } from '../version';
 
-export const GLOBAL_CONTEXT_MANAGER_API_KEY = Symbol.for(
-  'io.opentelemetry.js.api.context'
-);
-export const GLOBAL_METRICS_API_KEY = Symbol.for(
-  'io.opentelemetry.js.api.metrics'
-);
-export const GLOBAL_PROPAGATION_API_KEY = Symbol.for(
-  'io.opentelemetry.js.api.propagation'
-);
-export const GLOBAL_TRACE_API_KEY = Symbol.for('io.opentelemetry.js.api.trace');
+const _global = _globalThis as OTelGlobal;
+const acceptableRange = new semver.Range(`^${VERSION}`);
+const GLOBAL_OPENTELEMETRY_API_KEY = Symbol.for('io.opentelemetry.js.api');
 
-type Get<T> = (version: number) => T;
-type OtelGlobal = Partial<{
-  [GLOBAL_CONTEXT_MANAGER_API_KEY]: Get<ContextManager>;
-  [GLOBAL_METRICS_API_KEY]: Get<MeterProvider>;
-  [GLOBAL_PROPAGATION_API_KEY]: Get<TextMapPropagator>;
-  [GLOBAL_TRACE_API_KEY]: Get<TracerProvider>;
-}>;
+export function registerGlobal(type: 'trace', instance: TracerProvider): void;
+export function registerGlobal(type: 'metrics', instance: MeterProvider): void;
+export function registerGlobal(type: 'context', instance: ContextManager): void;
+export function registerGlobal(
+  type: 'propagation',
+  instance: TextMapPropagator
+): void;
+export function registerGlobal(type: keyof OTelGlobalApi, instance: any) {
```

Comment:
I think `registerGlobal<K extends keyof OTelGlobalApi, instance: OTelGlobalApi[K]["instance"]): void` could be used here instead manually writing down all overloads.
Maybe the `Partial` below needs to be move into an extra type.

Reply:
That's what I had before, but typescript made me add an `any` type because it didn't understand that if `type === "trace"` then `instance` is `TracerProvider`. It thinks it could be `ContextManager`, `TextMapPropagator`, etc:

```typescript
export function registerGlobal<Type extends keyof OTelGlobalApi>(type: Type, instance: ApiTypes[Type]): void {
  _global[GLOBAL_OPENTELEMETRY_API_KEY] =
    _global[GLOBAL_OPENTELEMETRY_API_KEY] ?? {};

  const api = _global[GLOBAL_OPENTELEMETRY_API_KEY]!;
  if(api[type]) {
    // already registered an API of this type
    return;
  }

  api[type] = {
    instance: instance as any,
    version: VERSION,
  };
}

```

Reply:
sad.
In case you decide to move the version into `OTelGlobalApi` and skip the `Signal<T>` type it should work.